libraries/command_line_tools.rb

```diff
@@ -1,34 +1,59 @@
 include Chef::Mixin::ShellOut
+require 'plist'
 
 module MacOS
   class CommandLineTools
     attr_reader :version
 
     def initialize
-      install_sentinel = '/tmp/.com.apple.dt.CommandLineTools.installondemand.in-progress'
-      FileUtils.touch install_sentinel
-      FileUtils.chown 'root', 'wheel', install_sentinel
+      if installed.empty?
+        enable_install_on_demand
+        @version = latest_from_catalog
+      else
+        @version = latest_installed
+      end
+    end
 
-      @version = if available.empty?
-                   'No Command Lines Tools available from Software Update Catalog!'
-                 elsif platform_specific.empty?
-                   "No Command Line Tools specific to #{macos_version} available from Software Update Catalog!"
-                 else
-                   latest.tr('*', '').strip
-                 end
+    def installed
+      packages = Plist.parse_xml(install_history_plist)
+      packages.select { |package| package['displayName'].match? 'Command Line Tools' }
     end
 
-    def latest
-      versions = platform_specific.map { |product| Gem::Version.new xcode_version(product) }
-      platform_specific.detect { |product| product.include? versions.max.version }
+    def latest_installed
+      [installed.last['displayName'], installed.last['displayVersion']].join '-'
+    end
+
+    def latest_from_catalog
+      if available.empty?
+        Chef::Log.warn 'No Command Line Tools available from Software Update Catalog!'
+      else
+        catalog_recommendation_label.tr('*', '').gsub('Label: ', '').strip
+      end
+    end
+
+    def catalog_recommendation_label
+      if platform_specific.empty?
+        Chef::Log.info "No Command Line Tools specific to #{macos_version} available from Software Update Catalog, selecting by maximum Xcode version."
+        versions = available.map { |product| Gem::Version.new xcode_version(product) }
+        available.detect { |product| product.include? versions.max.version }
+      else
+        versions = platform_specific.map { |product| Gem::Version.new xcode_version(product) }
+        platform_specific.detect { |product| product.include? versions.max.version }
```

Comment:
There's a little bit of duplication here – thinking we might be able to extract to a new method?

```ruby
def recommended_versions(software_update_package_list)
  versions = software_update_package_list.map { |product| Gem::Version.new xcode_version(product) }
  software_update_package_list.detect { |product| product.include? versions.max.version }
end
```

then:

```ruby
def catalog_recommendation_label
  unless platform_specific.empty?
    recommended_versions(available)
  else
    Chef::Log.info "No Command Line Tools specific to #{macos_version} available from Software Update Catalog"
    Chef::Log.info "Command Line Tools version will be selected using maximum Xcode version."
    recommended_versions(platform_specific)
  end
end
```

Reply:
Thanks - updated.